### PR TITLE
Feat/Big Panda Event Handler

### PR DESCRIPTION
**Pull request by rhajek**
Here is initial implementation of BigPanda event handler #2304. It is not tested with on production `https://api.bigpanda.io/data/v2/alerts` yet because we don't have BigPanda account.

Example of tick script:

```
stream
  |from()
    .measurement('cpu')
  |alert()
    .warn(lambda: "usage_idle" < 20)
    .stateChangesOnly()
    .message('Hey, check your CPU')
    .bigPanda()
```

kapacitor.conf

```
[bigpanda]
  enabled = true
  app-key = "my-app-key"
  token = "my-token-123"
  state-changes-only = false
  insecure-skip-verify = true
```

###### Required for all non-trivial PRs
- [ ] Rebased/mergable
- [ ] Tests pass
- [ ] CHANGELOG.md updated
- [ ] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

###### Required only if applicable
_You can erase any checkboxes below this note if they are not applicable to your Pull Request._
- [ ] [TICKscript Spec](https://github.com/influxdata/kapacitor/blob/master/tick/TICKscript.md) updated

### CODE REVIEW

File: services/alert/service.go [hunk 1]
```diff
@@ -763,17 +767,17 @@ func (s *Service) createHandlerFromSpec(spec HandlerSpec) (handler, error) {
 		keyvalue.KV("topic", spec.Topic),
 	}
 	switch spec.Kind {
-	case "aggregate":
-		c := newDefaultAggregateHandlerConfig(s.EventCollector)
```
Comment (docmerlin) [inline services/alert/service.go:767]: Er why was this removed?

**Review by docmerlin — CHANGES_REQUESTED** [review]
Can you ping me when it is passing tests?

File: services/alert/service.go (hunk 1 above)
Comment (rhajek) [reply services/alert/service.go]: This was mistake, thanks for pointing that!

**Comment by rhajek** [comment]
@docmerlin, tests are now passing in run 0 and 2 but I don't understand why is the TestService_GetConfig failing in run 1. On local machine TestService_GetConfig is also passing. What is the reason to have these 3 runs, I don't see any differences between them? Thanks.

CI link:
https://app.circleci.com/pipelines/github/influxdata/kapacitor/94/workflows/60e3e4e8-9c7c-462f-a57d-9fd1881fa285/jobs/4274/parallel-runs/1?filterBy=ALL

image

**Comment by rhajek** [comment]
It looks like flaky test, after rerun it is all green. Now I need real BigPanda testing account to verify implementation.

**Comment by docmerlin** [comment]
@rhajek working on getting us an account so we can do some manual testing.

**Review by docmerlin — CHANGES_REQUESTED** [review]
This is a marker so we can remember we need to test against actual BP.

**Comment by timhallinflux** [comment]
I'm working on getting us a test account with the BigPanda team.

**Comment by AlienHoboken** [comment]
Hey guys, couple quick comments on the default `kapacitor.conf` you proposed.

1. Do you expect any of the properties on the alert to change after the initial trigger (eg. `CRITICAL` status)? If not then I would suggest setting `state-changes-only = true`. BigPanda has [deduplication logic](https://docs.bigpanda.io/docs/alert-correlation-logic#event-deduping) which will drop these events if the timestamp hasn't changed (or merge if timestamp has but status hasn't) so there may not be value in repeatedly sending the alert, it'll appear as a single alert in BigPanda anyways.

2. I'd suggest setting `insecure-skip-verify = false`. Our API `https://api.bigpanda.io/data/v2/alerts` should have a valid TLS certificate unless something on your side necessitates this feature (proxy, etc).

**Comment by rhajek** [comment]
Hi @AlienHoboken 

1. `state-changes-only` should behave the same in all service integrations, common use case is that user specify deduplication directly in tick script.

```
stream
  |from()
    .measurement('cpu')
  |alert()
    .warn(lambda: "usage_idle" < 20)
    .stateChangesOnly()
    .message('Hey, check your CPU')
    .bigPanda()
```

2. I agree, `insecure-skip-verify = false` should be the default, insecure https is only for testing and non production use cases

Kapacitor Event can also contain map of custom tags and fields, that depends on data (user defined query). All additional tags are added to BP alert as `Additional attributes`, for additional fields I am not sure if they are important in BP, so they are not included.

**Comment by AlienHoboken** [comment]
@rhajek Thanks, #1 makes sense to me 👍 

Awesome that you are sending the additional tags. For additional fields could you possibly share an example of what data that might contain? I could advise if it makes sense to send also.

**Comment by rhajek** [comment]
First kapacitor alert in bigpanda:

### kapacitor.conf:
```
[bigpanda]
  enabled = true
  app-key = "8f57...e4"
  token = "1d9...cec"
```
### tickscript:
```
stream
  |from()
    .measurement('cpu')
  |alert()
    .warn(lambda: "usage_idle" < 20)
    .stateChangesOnly()
    .message('Hey, check your CPU')
    .bigPanda()
``` 

### bigpanda ui
image

### For cpu monitoring example kapacitor alert data looks like this:
image

**Comment by timhallinflux** [comment]
any other changes requested @AlienHoboken or @docmerlin ?  Otherwise, let's also land this for 1.5.8...

File: services/bigpanda/service.go [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+	}
+
+}
+
+func (s *Service) Test(options interface{}) error {
+	o, ok := options.(*testOptions)
+	if !ok {
+		return fmt.Errorf("unexpected options type %T", options)
+	}
+	//c := s.config()
```
Comment (docmerlin) [inline services/bigpanda/service.go:116]: delete this comment

**Review by docmerlin — CHANGES_REQUESTED** [review]
This looks good, only really one thing to clean up (the forgotten commented out code).  Please fix it before merging.

File: services/bigpanda/service.go (hunk 2 above)
Comment (rhajek) [reply services/bigpanda/service.go]: fixed

**Review by docmerlin — APPROVED** [review]

**Comment by AlienHoboken** [comment]
Looks great to me @timhallinflux . It is possible customers _might_ want to include some information from the fields section, not only tags, in their alerts but we'll see that in feedback probably.
